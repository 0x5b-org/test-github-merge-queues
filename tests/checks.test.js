import { Octokit, RequestError } from "octokit";
import { beforeAll, describe, test } from "vitest";
import _ from 'lodash';
import 'lodash.product';
import { createAppAuth } from "@octokit/auth-app";

const octokit = new Octokit({
  authStrategy: createAppAuth,
  auth: {
    appId: 1178750,
    privateKey: process.env.GITHUB_APP_PRIVATE_KEY,
    installationId: 62652949
  }
});

console.log(await octokit.rest.apps.getAuthenticated());

const statuses = [
  'queued',
  'in_progress',
  'completed',
  // Only GitHub Actions can set a status of waiting, pending, or requested.
  // 'waiting',
  // 'requested',
  // 'pending',
];

const conclusions = [
  'action_required',
  'cancelled',
  'failure',
  'neutral',
  'success',
  'skipped',
  // You cannot change a check run conclusion to stale, only GitHub can set this.
  // 'stale', 
  'timed_out',
];

const { data: main } = await octokit.rest.git.getRef({
  owner: 'jonathanmorley',
  repo: 'repository-config-testbed',
  ref: 'heads/main'
});

const rulesets = await octokit.paginate(octokit.rest.repos.getRepoRulesets, {
  owner: 'jonathanmorley',
  repo: 'repository-config-testbed',
  includes_parents: false
});

const branches = await octokit.paginate(octokit.rest.repos.listBranches, {
  owner: 'jonathanmorley',
  repo: 'repository-config-testbed'
});

let featureBranchSha;

// Ensure feature branch
beforeAll(async () => {
  const { data: mainTree } = await octokit.rest.git.getTree({
    owner: 'jonathanmorley',
    repo: 'repository-config-testbed',
    tree_sha: main.object.sha
  });

  const { data: tree } = await octokit.rest.git.createTree({
    owner: 'jonathanmorley',
    repo: 'repository-config-testbed',
    base_tree: mainTree.sha,
    tree: [
      {
        path: 'test_file',
        content: Buffer.from('Hello World!').toString('base64'),
        mode: '100644',
      }
    ]
  });

  const { data: commit } = await octokit.rest.git.createCommit({
    owner: 'jonathanmorley',
    repo: 'repository-config-testbed',
    message: 'Create feature branch',
    tree: tree.sha,
    parents: [main.object.sha]
  });

  if (branches.find(branch => branch.name === 'checks/feature')) {
    await octokit.rest.git.updateRef({
      owner: 'jonathanmorley',
      repo: 'repository-config-testbed',
      ref: 'heads/checks/feature',
      sha: commit.sha,
      force: true
    })
  } else {
    await octokit.rest.git.createRef({
      owner: 'jonathanmorley',
      repo: 'repository-config-testbed',
      ref: 'refs/heads/checks/feature',
      sha: commit.sha
    });
  }

  featureBranchSha = commit.sha;
})

describe.concurrent.for(_.product(statuses, conclusions))('Check %s, %s', async ([status, conclusion]) => {
  // Cleanup
  beforeAll(async ({ }) => {
    // Delete branch
    const branch = branches.find(branch => branch.name === `checks/${status}/${conclusion}/main`);
    if (branch) {
      await octokit.rest.git.deleteRef({
        owner: 'jonathanmorley',
        repo: 'repository-config-testbed',
        ref: `heads/${branch.name}`
      });
    }
  }, 90_000);

  // Setup
  beforeAll(async () => {
    // Upsert ruleset
    const ruleset = {
      owner: 'jonathanmorley',
      repo: 'repository-config-testbed',
      name: `Checks ${status} ${conclusion}`,
      target: 'branch',
      enforcement: 'active',
      conditions: {
        ref_name: {
          include: [`refs/heads/checks/${status}/${conclusion}/main`],
          exclude: []
        }
      },
      rules: [
        {
          type: 'required_status_checks',
          parameters: {
            strict_required_status_checks_policy: false,
            do_not_enforce_on_create: true,
            required_status_checks: [
              {
                context: `${status}/${conclusion}`,
                integration_id: 1178750
              }
            ]
          }
        }
      ]
    };

    const rulesetId = rulesets.find(r => r.name === ruleset.name)?.id;
    if (rulesetId) await octokit.rest.repos.updateRepoRuleset({ ...ruleset, ruleset_id: rulesetId });
    else await octokit.rest.repos.createRepoRuleset(ruleset);
  
    // Create branch
    await octokit.rest.git.createRef({
      owner: 'jonathanmorley',
      repo: 'repository-config-testbed',
      ref: `refs/heads/checks/${status}/${conclusion}/main`,
      sha: main.object.sha
    });

    // Create check on feature branch
    await octokit.rest.checks.create({
      owner: 'jonathanmorley',
      repo: 'repository-config-testbed',
      head_sha: featureBranchSha,
      name: `${status}/${conclusion}`,
      status,
      conclusion
    });
  }, 40_000);

  test('force update the branch', async ({ expect }) => {
    let result = 'success';
    
    try {
      await octokit.rest.git.updateRef({
        owner: 'jonathanmorley',
        repo: 'repository-config-testbed',
        ref: `heads/checks/${status}/${conclusion}/main`,
        sha: featureBranchSha,
        force: true
      });
    } catch (err) {
      expect(err).toBeInstanceOf(RequestError);
      if (err instanceof RequestError) result = err.message;
    }

    expect(result).toMatchSnapshot();
  });
});
