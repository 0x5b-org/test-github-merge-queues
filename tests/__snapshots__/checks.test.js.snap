// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Check completed, action_required > force update the branch 1`] = `
"Repository rule violations found

Required status check "completed/action_required" is action required.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check completed, cancelled > force update the branch 1`] = `
"Repository rule violations found

Required status check "completed/cancelled" is cancelled.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check completed, failure > force update the branch 1`] = `
"Repository rule violations found

Required status check "completed/failure" is failing.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check completed, neutral > force update the branch 1`] = `"success"`;

exports[`Check completed, skipped > force update the branch 1`] = `"success"`;

exports[`Check completed, success > force update the branch 1`] = `"success"`;

exports[`Check completed, timed_out > force update the branch 1`] = `
"Repository rule violations found

Required status check "completed/timed_out" is timed out.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check in_progress, action_required > force update the branch 1`] = `
"Repository rule violations found

Required status check "in_progress/action_required" is action required.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check in_progress, cancelled > force update the branch 1`] = `
"Repository rule violations found

Required status check "in_progress/cancelled" is cancelled.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check in_progress, failure > force update the branch 1`] = `
"Repository rule violations found

Required status check "in_progress/failure" is failing.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check in_progress, neutral > force update the branch 1`] = `"success"`;

exports[`Check in_progress, skipped > force update the branch 1`] = `"success"`;

exports[`Check in_progress, success > force update the branch 1`] = `"success"`;

exports[`Check in_progress, timed_out > force update the branch 1`] = `
"Repository rule violations found

Required status check "in_progress/timed_out" is timed out.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check queued, action_required > force update the branch 1`] = `
"Repository rule violations found

Required status check "queued/action_required" is action required.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check queued, cancelled > force update the branch 1`] = `
"Repository rule violations found

Required status check "queued/cancelled" is cancelled.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check queued, failure > force update the branch 1`] = `
"Repository rule violations found

Required status check "queued/failure" is failing.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;

exports[`Check queued, neutral > force update the branch 1`] = `"success"`;

exports[`Check queued, skipped > force update the branch 1`] = `"success"`;

exports[`Check queued, success > force update the branch 1`] = `"success"`;

exports[`Check queued, timed_out > force update the branch 1`] = `
"Repository rule violations found

Required status check "queued/timed_out" is timed out.

 - https://docs.github.com/rest/git/refs#update-a-reference"
`;
